bftengine/src/bftengine/IncomingMsgsStorageImp.cpp

```diff
@@ -42,7 +48,14 @@ IncomingMsgsStorageImp::~IncomingMsgsStorageImp() {
 void IncomingMsgsStorageImp::start() {
   if (!dispatcherThread_.joinable()) {
     std::future<void> futureObj = signalStarted_.get_future();
-    dispatcherThread_ = std::thread([=] { dispatchMessages(signalStarted_); });
+    dispatcherThread_ = std::thread([=] {
+#ifdef USE_LOG4CPP
```

Comment:
I = dislike the use of conditional code for logging here. If we start doing this, it will litter the code everywhere. We've already removed it in a bunch of places prior to this. 

While I'm fine with it for now, as it definitely helps debugging in the short term, we should really talk about cleaning up logging altogether.  My suggestion is just to force the use of LOG4CPP. We bend over backwards to prevent having a dependency for logging, but it essentially means we have no straightforward way to even configure our log levels for tests. 

If we decide to maintain an abstraction, we should make it much more advanced, provide an interface, and allow the user of the library to pass in the implementation of the logger themselves. 
We should really consider a

Reply:
Hi Andrew, thanks for bringing this up! I agree the current situation is not great, and suggest the following:

1) **[Short term]** as a first step, we should always use LOG4CPP for CI & local testing. This way we won't be stuck because our logging isn't flexible enough.

2) **[Long term]** In terms of general logging approach, I don't think that forcing the use of LOG4CPP is a good idea. Concord BFT is a library, so it should provide a thin interface where different logging frameworks can be plugged. It seems like a bigger and more complex effort than step 1.

Both of those items are added to our backlog. The first one can be executed as part of Concord-BFT testing. The second one seems like typical Concord-BFT tech debt, which should be prioritized with respect to the needs of Concord BFT library clients.

Best regards,
Teo